File: HomeAutomation/Code/Controllers/VoiceCommandSystem.java
package Code.Controllers;

import Code.Interfaces.Component;
import Code.StateManagementClasses.StateManager;
import java.util.ArrayList;
import java.util.List;

public class VoiceCommandSystem implements Component {
    private StateManager stateManager;
    private List<String> commandHistory = new ArrayList<>();

    public VoiceCommandSystem(StateManager stateManager) {
        this.stateManager = stateManager;
    }

    @Override
    public void onStateChange(String key, Object value) {
        if ("voiceCommand".equals(key) && value instanceof String) {
            processVoiceCommand((String) value);
        }
    }

    private void processVoiceCommand(String command) {
        commandHistory.add(command);
        System.out.println("Processing voice command: " + command);

        switch (command.toLowerCase()) {
            case "turn on the lights":
                stateManager.updateState("lights", "on");
                System.out.println("Lights turned ON.");
                break;

            case "turn off the lights":
                stateManager.updateState("lights", "off");
                System.out.println("Lights turned OFF.");
                break;

            case "increase temperature":
                stateManager.updateState("temperature", "increase");
                System.out.println("Temperature increased.");
                break;

            case "decrease temperature":
                stateManager.updateState("temperature", "decrease");
                System.out.println("Temperature decreased.");
                break;

            default:
                System.out.println("Unknown voice command: " + command);
        }
    }

    public List<String> getCommandHistory() {
        return commandHistory;
    }
}

File: HomeAutomation/HomeAutomation/Componenets/VoiceCommandSystem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace HomeAutomation.Componenets
{
    public class VoiceCommandSystem
    {
        private Dictionary<string, Action> commands = new();

        public VoiceCommandSystem()
        {
            commands["turn on lights"] = () => Console.WriteLine("Lights turned ON");
            commands["turn off lights"] = () => Console.WriteLine("Lights turned OFF");
            commands["increase temperature"] = () => Console.WriteLine("Temperature increased");
        }

        public void ProcessCommand(string command)
        {
            if (commands.ContainsKey(command))
            {
                commands[command].Invoke();
            }
            else
            {
                Console.WriteLine("Unknown command.");
            }
        }
    }

}
